Signal/src/view controllers/FullImageViewController.m

```diff
@@ -128,56 +159,106 @@ - (void)populateImageView:(UIImage *)image {
-        zoom = CGSizeMake(self.view.bounds.size.width * self.scrollView.zoomScale,
-                          self.view.bounds.size.height * self.scrollView.zoomScale);
-        targetZoomRect = CGRectMake(
-            convertCoord.x - (zoom.width / 2.0f), convertCoord.y - (zoom.height / 2.0f), zoom.width, zoom.height);
-        targetInsets = [self contentInsetForScrollView:1.0f];
+        [self.view becomeFirstResponder];
+        
+        if ([UIMenuController sharedMenuController].isMenuVisible) {
+            [[UIMenuController sharedMenuController] setMenuVisible:NO
+                                                           animated:NO];
+        }
+
+        NSArray *menuItems = @[
+                               [[UIMenuItem alloc] initWithTitle:NSLocalizedString(@"ATTACHMENT_VIEW_COPY_ACTION", @"Short name for edit menu item to copy contents of media message.")
+                                                          action:@selector(copyAttachment:)],
+                               [[UIMenuItem alloc] initWithTitle:NSLocalizedString(@"ATTACHMENT_VIEW_SAVE_ACTION", @"Short name for edit menu item to save contents of media message.")
+                                                          action:@selector(saveAttachment:)],
+                               // TODO: We should implement sharing.
+                               //                                                              [[UIMenuItem alloc] initWithTitle:NSLocalizedString(@"ATTACHMENT_VIEW_SHARE_ACTION", @"Short name for edit menu item to share contents of media message.")
+                               //                                                                                         action:@selector(shareAttachment:)],
+                               ];
+        [UIMenuController sharedMenuController].menuItems = menuItems;
+        CGPoint location = [sender locationInView:self.view];
+        CGRect targetRect = CGRectMake(location.x,
+                                       location.y,
+                                       1, 1);
+        [[UIMenuController sharedMenuController] setTargetRect:targetRect
+                                                        inView:self.view];
+        [[UIMenuController sharedMenuController] setMenuVisible:YES
+                                                       animated:YES];
     }
+}
 
-    self.view.userInteractionEnabled = NO;
+- (void)performEditingActionWithSelector:(SEL)selector {
+    OWSAssert(self.messageItem.messageType == TSIncomingMessageAdapter ||
+              self.messageItem.messageType == TSOutgoingMessageAdapter);
+    OWSAssert([self.messageItem isMediaMessage]);
+    OWSAssert([self.messageItem isKindOfClass:[TSMessageAdapter class]]);
+    OWSAssert([self.messageItem conformsToProtocol:@protocol(OWSMessageEditing)]);
+    OWSAssert([[self.messageItem media] isKindOfClass:[TSPhotoAdapter class]] ||
+              [[self.messageItem media] isKindOfClass:[TSAnimatedAdapter class]]);
```

Comment:
No, one thing I'd like to do is generalize `FullImageViewController` to a `MediaViewController` or `AttachmentViewController` and use it for audio and video as well.  But I haven't done that yet.